distributed/scheduler.py

```diff
@@ -654,6 +658,69 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict[str, Any]:
             members=True,
         )
 
+    def add_to_processing(self, ts: TaskState) -> None:
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        tg = ts.group
+        self.task_groups_count[tg.name] += 1
+        scheduler.task_groups_count_global[tg.name] += 1
+        self.processing.add(ts)
+        for dts in ts.dependencies:
+            if self not in dts.who_has:
+                self.need_replica(dts)
+
+    # TODO: Scheduler.add_replica should call this
+    def need_replica(self, ts: TaskState) -> None:
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        if ts not in self.needs_what:
+            self.needs_what[ts] = 1
+            nbytes = ts.get_nbytes()
+            self._network_occ += nbytes
+            scheduler._network_occ_global += nbytes
+        else:
+            self.needs_what[ts] += 1
+
+    def dont_need_replica(self, ts: TaskState) -> None:
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        if ts in self.needs_what:
+            self.needs_what[ts] -= 1
+            if not self.needs_what[ts]:
+                del self.needs_what[ts]
+                nbytes = ts.get_nbytes()
+                self._network_occ -= nbytes
+                scheduler._network_occ_global -= nbytes
+
+    def remove_from_processing(self, ts: TaskState) -> None:
+        if ts not in self.processing:
+            return
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        self.task_groups_count[ts.group.name] -= 1
+        if not self.task_groups_count[ts.group.name]:
+            del self.task_groups_count[ts.group.name]
+        scheduler.task_groups_count_global[ts.group.name] -= 1
+        if not scheduler.task_groups_count_global[ts.group.name]:
+            del scheduler.task_groups_count_global[ts.group.name]
+        self.processing.remove(ts)
+        # TODO Loop over the smaller set of the two, see Scheduler.get_comm_cost
+        for dts in ts.dependencies:
+            if dts in self.needs_what:
+                self.dont_need_replica(dts)
+
+    @property
+    def occupancy(self):
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        res = 0
+        for group_name, count in self.task_groups_count.items():
+            # TODO: Deal with unknown tasks better
+            prefix = scheduler.task_groups[group_name].prefix
+            assert prefix
```

Comment:
just added this assert for mypy and didn't think about how it is implemented. Since we didn't define a `__bool__` this iterates over the groups and prefixes. That's obviously way too expensive if we're doing it in this for loop

https://github.com/dask/distributed/blob/1fd07f03cacee6fde81d13282568a727bce789b9/distributed/scheduler.py#L814-L815